feat: 마이 페이지 기본 기능 구현

djlim00 opened the pull request:
## 📚 개요

- 마이 페이지 기본 기능을 구현했습니다.

## 🕐 리뷰 예상 시간

> 20m

## ❗❗ 중요한 변경점(Option)

- MyPage 화면에서 내가 쓴 글과, 내가 좋요을 누른 게시물을 가져오는 api 구현
- Service단 함수명을 더 명확하게 수정
- MayPage에서 닉네임 수정, 기본정보 수정, 부가정보 수정 기능구현

djlim00 committed:
feat: 내가 쓴 게시물, 좋아요 누른 게시물 조회 기능 구현

- MyPage 화면에서 내가 쓴 글과, 내가 좋요을 누른 게시물을 가져오는 api 구현
- Service단 함수명을 더 명확하게 수정

djlim00 committed:
feat: 닉네임 수정 구현

- 닉네임 수정을 구현했습니다.

djlim00 committed:
feat: 마이 페이지 정보 수정 기능 구현

- 마이페이지 기본정보(거주지, 나이, 성별)을 수정하는 기능 추가
- 마이페이지 기본정보(경력, 관심직종)을 수정하는 기능 추가
- 닉네임 수정을 User -> MyPage로 이동

djlim00 committed:
fix: MyPageService 오타 수정

djlim00 commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/service/MyPageService.kt:
```diff
@@ -0,0 +1,102 @@
+                    user = user,
+                    category1 = interests.getOrNull(0),
+                    category2 = interests.getOrNull(0),
+                    category3 = interests.getOrNull(0),
```
Thx!

djlim00 committed:
fix: MyPageController 오타 수정

-일관되지 못하게 작성된 함수명 수정

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/controller/MyPageController.kt:
```diff
@@ -0,0 +1,65 @@
+    )
+    @GetMapping("/articles/{type}")
+    fun getMyArticles(
+        @PathVariable type: String,
```
type같은 거는 보통 enum을 만들어두는게 좋습니다.

그리고 PathVariable보다는 query로 받는게 더 좋아보이긴 하는데 어떻게 생각하시나요

djlim00 replied:
저는 쿼리는 값을 받아오기 위해서 보통 사용하고 위와 같이 기능을 구분하기 위해서 쿼리를 사용하길 꺼려하는데, 
솔직히 쿼리가 더 편하긴 할거 같아요. 반영하겠습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/controller/MyPageController.kt:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Operation(summary = "닉네임 수정", description = "마이페이지에서 닉네임을 수정합니다.")
+    @PatchMapping("/nickname/update")
```
```suggestion
    @PatchMapping("/nickname")
```

rest에서는 url에 action이 들어가면 안됩니다. action은 http method로 표현해주세요

djlim00 replied:
반영하겠습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/global/enums/CustomError.kt:
```diff
@@ -7,6 +7,7 @@ enum class CustomError(val status: HttpStatus, val message: String) {
     // common
     ACCESS_TOKEN_EXPIRED(UNAUTHORIZED, "Access Token이 만료되었습니다."),
     INVALID_TOKEN(UNAUTHORIZED, "유효하지 않은 토큰입니다."),
+    INVALID_URL(NOT_FOUND, "잘못된 요청 URL입니다."),
```
위에 리뷰 반영하시면 이건 필요없어질거에요

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/service/MyPageService.kt:
```diff
@@ -0,0 +1,102 @@
+            val commentCount = article.comments.size
+            val uploadAt = articleService.getPreviewUploadTime(article.createdAt)
+
+            ArticlePreviewResponse(
+                articleId = article.id,
+                title = article.title,
+                text = article.text,
+                likeCount = article.likeCount,
+                commentCount = commentCount,
+                uploadAt = uploadAt,
+                thumbnailImage = article.images.firstOrNull()?.url,
+                articleKind = article.articleKindId,
+                isLiked = likedArticleIds.contains(article.id),
+            )
```
개인적으로 이렇게 직접 할당하는 것보다 
객체 내부에 static of를 만들어서 할당하는게 좀 더 깔끔해서 좋아합니다.

우리 프로젝트에도 있을 테니 참고해보세요

djlim00 replied:
오호.. 좋은거 같네요 반영하겠습니다

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/controller/MyPageController.kt:
```diff
@@ -0,0 +1,65 @@
+    @PatchMapping("/basic-info")
+    fun updateBasicInfo(
+        @RequestUser user: User,
+        @RequestBody request: UpdateBasicInfoResponse
```
```suggestion
        @RequestBody request: UpdateBasicInfoRequest
```

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/controller/MyPageController.kt:
```diff
@@ -0,0 +1,65 @@
+    @PatchMapping("/sub-info")
+    fun updateSubInfo(
+        @RequestUser user: User,
+        @RequestBody request: UpdateSubInfoResponse
```
```suggestion
        @RequestBody request: UpdateSubInfoRequest
```

djlim00 replied:
민망하네요.. 수정하겠습니다.

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/service/MyPageService.kt:
```diff
@@ -0,0 +1,102 @@
+        val likedArticleIds = articleLikeService.findLikedArticleIds(user.id, articleIds)
+
+        return articles.map { article ->
+            val commentCount = article.comments.size
+            val uploadAt = articleService.getPreviewUploadTime(article.createdAt)
```
commentCount는 article이 들고있어야 할 것 같구요 (코멘트 개수 변경될 때 업데이트하게끔)
uploadAt은 진짜로 이건 프론트가 하는게 맞아요

djlim00 replied:
아 commentCount는 Article 쪽에서 복붙하다 남은 잔재네요ㅠ
uploadAt도 그냥 createdAt 넘기는 걸로 수정했습니다.

djlim00 committed:
refactor: Article dto 반환방식 변경

- 객체 내부에 static of를 만들어서 할당하였습니다.

djlim00 committed:
refactor: Auth dto 반환방식 변경

- 객체 내부에 static of를 만들어서 할당하였습니다.

djlim00 committed:
refactor: Comment dto 반환방식 변경

- 객체 내부에 static of를 만들어서 할당하였습니다.

djlim00 committed:
refactor: User dto 반환방식 변경

- 객체 내부에 static of를 만들어서 할당하였습니다.

djlim00 committed:
refactor: uploadAt -> createAt 반환 변경

- 업로드 시점을 알려주는 역할을 프론트에서 하도록 했습니다.

djlim00 committed:
update: MyPage 도메인 수정

- mypage/articles 엔드 포인트 쿼리 파라미터로 변경
- 파라미터 값 enum으로 정의
- 오타 수정
- dto 반환 방식 static of으로 변경
- 필요없어진 Custom Error 삭제

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/dto/AuthorizeResponse.kt:
```diff
@@ -1,8 +1,26 @@
-)
\ No newline at end of file
+) {
+    companion object {
+        fun of(
+            user: User,
+            accessToken: String,
+            refreshToken: String,
+            isRegistered: Boolean
+        ): AuthorizeResponse {
+            return AuthorizeResponse(
+                userId = user.id,
+                accessToken = accessToken,
+                refreshToken = refreshToken,
+                isRegistered = isRegistered
+            )
+        }
+    }
```
이 정도면 사실 안만들어도 돼요

djlim00 replied:
아ㅋㅋㅋㅋ 처음 써봐서 연습하는 겸 진짜 다 만들어서 필요없는거 지울게요. 제가 생각해도 에바긴하네요

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/dto/RegisterResponse.kt:
```diff
@@ -1,6 +1,17 @@
-)
\ No newline at end of file
+) {
+    companion object {
+        fun of(isRegistered: Boolean, user: User): RegisterResponse {
+            return RegisterResponse(
+                isRegistered = isRegistered,
+                userId = user.id
+            )
+        }
+    }
```
여기도 안만들어도 괜찮습니다

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/auth/dto/jwt/ReissueAccessTokenResponse.kt:
```diff
@@ -2,4 +2,12 @@ package com.arabyte.arabyteapi.domain.auth.dto.jwt
     val accessToken: String
-)
+) {
+    companion object {
+        fun of(accessToken: String): ReissueAccessTokenResponse {
+            return ReissueAccessTokenResponse(
+                accessToken = accessToken
+            )
+        }
```
여기도요

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/controller/MyPageController.kt:
```diff
@@ -0,0 +1,66 @@
+    )
+    @GetMapping("/articles")
+    fun getMyArticles(
+        @RequestParam type: MyPageArticleType,
+        @RequestUser user: User,
```
```suggestion
        @RequestUser user: User,
        @RequestParam type: MyPageArticleType,
```

request user를 항상 맨위에 쓰는걸로하죠

djlim00 replied:
반영하겠습니다~

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/controller/MyPageController.kt:
```diff
@@ -0,0 +1,66 @@
+) {
+    @Operation(
+        summary = "마이페이지 게시물 조회",
+        description = "?type=MY로 요청하면 내가 쓴 게시물, ?type=LIKE로 요청하면 내가 좋아요를 누른 게시물을 반환합니다."
```
굳이 쿼리 스트링을 사용하는 법까지 적을 필요는 없을 것 같구요
각 타입이 어떤걸 반환하지는만 적어줘도 충분할 것 같습니다.

djlim00 replied:
넵!

Ji-InPark commented on src/main/kotlin/com/arabyte/arabyteapi/domain/mypage/service/MyPageService.kt:
```diff
@@ -0,0 +1,83 @@
+        user.experienceYears = request.experienceYears
+        user.experienceMonths = request.experienceMonths
+
+        if (request.jobInterests.isNotEmpty()) {
+            val interests = request.jobInterests
+            val userJobInterests = user.jobInterests
+
+            if (userJobInterests != null) {
+                userJobInterests.category1 = interests.getOrNull(0)
+                userJobInterests.category2 = interests.getOrNull(1)
+                userJobInterests.category3 = interests.getOrNull(2)
+            } else {
+                val interestEntity = UserJobInterest(
+                    user = user,
+                    category1 = interests.getOrNull(0),
+                    category2 = interests.getOrNull(1),
+                    category3 = interests.getOrNull(2),
+                )
+                user.jobInterests = interestEntity
+            }
+        }
```
if문 depth는 최대한 줄여주세요.

그냥 어느 함수든 최대 depth는 1까지라고 생각하셔도 됩니다.

djlim00 committed:
refactor: 불필요한 of 삭제, if문 depth 수정

- 불필요한 of 메서드 삭제
- MyPageService의 updateSubInfo의 중첩 if문 수정
